File: backend/app/schemas/recopilacion.py
"""
Schemas Pydantic para Recopilacion Guiada (Fase 3).
"""
from datetime import datetime
from enum import Enum
from typing import Optional, List, Dict, Any, Union
from pydantic import BaseModel, Field, validator


# =============================================================================
# ENUMS
# =============================================================================

class EstadoSeccion(str, Enum):
    """Estados posibles de una seccion del EIA."""
    PENDIENTE = "pendiente"
    EN_PROGRESO = "en_progreso"
    COMPLETADO = "completado"
    VALIDADO = "validado"


class TipoRespuesta(str, Enum):
    """Tipos de respuesta para preguntas."""
    TEXTO = "texto"
    NUMERO = "numero"
    FECHA = "fecha"
    SELECCION = "seleccion"
    SELECCION_MULTIPLE = "seleccion_multiple"
    ARCHIVO = "archivo"
    COORDENADAS = "coordenadas"
    TABLA = "tabla"
    BOOLEANO = "booleano"


class SeveridadInconsistencia(str, Enum):
    """Niveles de severidad."""
    ERROR = "error"
    WARNING = "warning"
    INFO = "info"


class EstadoExtraccion(str, Enum):
    """Estados de extraccion."""
    PENDIENTE = "pendiente"
    PROCESANDO = "procesando"
    COMPLETADO = "completado"
    ERROR = "error"


# =============================================================================
# SCHEMAS DE PREGUNTAS
# =============================================================================

class OpcionSeleccion(BaseModel):
    """Opcion para preguntas de tipo seleccion."""
    valor: str
    etiqueta: str
    descripcion: Optional[str] = None


class ValidacionPregunta(BaseModel):
    """Reglas de validacion para una pregunta."""
    requerido: bool = True
    min: Optional[Union[int, float]] = None
    max: Optional[Union[int, float]] = None
    patron: Optional[str] = None
    mensaje_patron: Optional[str] = None


class PreguntaRecopilacionBase(BaseModel):
    """Schema base para preguntas de recopilacion."""
    capitulo_numero: int
    seccion_codigo: str
    seccion_nombre: Optional[str] = None
    codigo_pregunta: str
    pregunta: str
    descripcion: Optional[str] = None
    tipo_respuesta: TipoRespuesta = TipoRespuesta.TEXTO
    opciones: Optional[List[OpcionSeleccion]] = None
    valor_por_defecto: Optional[Any] = None
    validaciones: Optional[ValidacionPregunta] = None
    es_obligatoria: bool = True
    orden: int = 0


class PreguntaRecopilacionResponse(PreguntaRecopilacionBase):
    """Respuesta de pregunta de recopilacion."""
    id: int
    condicion_activacion: Optional[Dict[str, Any]] = None
    activo: bool = True

    class Config:
        from_attributes = True


class PreguntaConRespuesta(PreguntaRecopilacionResponse):
    """Pregunta con su respuesta actual."""
    respuesta_actual: Optional[Any] = None
    es_valida: bool = True
    mensaje_validacion: Optional[str] = None


# =============================================================================
# SCHEMAS DE SECCIONES
# =============================================================================

class SeccionBase(BaseModel):
    """Schema base para secciones."""
    capitulo_numero: int
    seccion_codigo: str
    seccion_nombre: Optional[str] = None


class SeccionConPreguntas(SeccionBase):
    """Seccion con sus preguntas."""
    preguntas: List[PreguntaConRespuesta] = []
    total_preguntas: int = 0
    preguntas_obligatorias: int = 0
    preguntas_respondidas: int = 0
    progreso: int = 0
    estado: EstadoSeccion = EstadoSeccion.PENDIENTE


class SeccionResumen(SeccionBase):
    """Resumen de estado de una seccion."""
    estado: EstadoSeccion
    progreso: int
    tiene_inconsistencias: bool = False
    documentos_vinculados: int = 0


# =============================================================================
# SCHEMAS DE CONTENIDO
# =============================================================================

class ContenidoEIABase(BaseModel):
    """Schema base para contenido EIA."""
    capitulo_numero: int
    seccion_codigo: str
    seccion_nombre: Optional[str] = None


class ContenidoEIACreate(ContenidoEIABase):
    """Schema para crear contenido EIA."""
    contenido: Dict[str, Any] = Field(default_factory=dict)


class ContenidoEIAUpdate(BaseModel):
    """Schema para actualizar contenido EIA."""
    contenido: Optional[Dict[str, Any]] = None
    estado: Optional[EstadoSeccion] = None
    notas: Optional[str] = None


class RespuestaIndividual(BaseModel):
    """Schema para guardar una respuesta individual."""
    codigo_pregunta: str
    valor: Any


class GuardarRespuestasRequest(BaseModel):
    """Request para guardar multiples respuestas."""
    respuestas: List[RespuestaIndividual]


class ContenidoEIAResponse(ContenidoEIABase):
    """Respuesta de contenido EIA."""
    id: int
    proyecto_id: int
    contenido: Dict[str, Any] = Field(default_factory=dict)
    estado: EstadoSeccion
    progreso: int
    documentos_vinculados: List[int] = []
    validaciones: List[Dict[str, Any]] = []
    inconsistencias: List[Dict[str, Any]] = []
    sugerencias: List[Dict[str, Any]] = []
    notas: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# =============================================================================
# SCHEMAS DE INCONSISTENCIAS
# =============================================================================

class InconsistenciaDetectada(BaseModel):
    """Inconsistencia detectada entre secciones."""
    regla_codigo: str
    regla_nombre: str
    capitulo_origen: int
    seccion_origen: str
    campo_origen: str
    valor_origen: Optional[Any] = None
    capitulo_destino: int
    seccion_destino: str
    campo_destino: str
    valor_destino: Optional[Any] = None
    mensaje: str
    severidad: SeveridadInconsistencia
    fecha_deteccion: Optional[datetime] = None


class ValidacionConsistenciaResponse(BaseModel):
    """Resultado de validacion de consistencia."""
    proyecto_id: int
    total_reglas_evaluadas: int
    inconsistencias: List[InconsistenciaDetectada]
    errores: int = 0
    warnings: int = 0
    es_consistente: bool

    @validator('es_consistente', always=True, pre=True)
    def calcular_consistencia(cls, v, values):
        return values.get('errores', 0) == 0


# =============================================================================
# SCHEMAS DE CAPITULOS
# =============================================================================

class CapituloProgreso(BaseModel):
    """Progreso de un capitulo."""
    numero: int
    titulo: str
    total_secciones: int
    secciones_completadas: int
    progreso: int
    estado: EstadoSeccion
    tiene_inconsistencias: bool = False
    secciones: List[SeccionResumen] = []


class ProgresoGeneralEIA(BaseModel):
    """Progreso general del EIA."""
    proyecto_id: int
    total_capitulos: int
    capitulos_completados: int
    progreso_general: int
    total_inconsistencias: int
    capitulos: List[CapituloProgreso]


# =============================================================================
# SCHEMAS DE EXTRACCION
# =============================================================================

class MapeoSeccionSugerido(BaseModel):
    """Mapeo sugerido de dato extraido a seccion."""
    capitulo_numero: int
    seccion_codigo: str
    campo: str
    valor: Any
    confianza: float = Field(ge=0.0, le=1.0)


class ExtraccionDocumentoBase(BaseModel):
    """Schema base para extraccion de documento."""
    documento_id: int
    tipo_documento: Optional[str] = None


class ExtraccionDocumentoRequest(ExtraccionDocumentoBase):
    """Request para extraer datos de documento."""
    forzar_reprocesar: bool = False


class ExtraccionDocumentoResponse(ExtraccionDocumentoBase):
    """Respuesta de extraccion de documento."""
    id: int
    proyecto_id: int
    datos_extraidos: Dict[str, Any] = Field(default_factory=dict)
    mapeo_secciones: List[MapeoSeccionSugerido] = []
    confianza_extraccion: float = 0.0
    estado: EstadoExtraccion
    errores: List[Dict[str, Any]] = []
    procesado_por: str = "claude_vision"
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class AplicarExtraccionRequest(BaseModel):
    """Request para aplicar datos extraidos a secciones."""
    extraccion_id: int
    mapeos_confirmados: List[MapeoSeccionSugerido]


# =============================================================================
# SCHEMAS DE VINCULACION DE DOCUMENTOS
# =============================================================================

class VincularDocumentoRequest(BaseModel):
    """Request para vincular documento a seccion."""
    documento_id: int
    capitulo_numero: int
    seccion_codigo: str


class DesvincularDocumentoRequest(BaseModel):
    """Request para desvincular documento de seccion."""
    documento_id: int
    capitulo_numero: int
    seccion_codigo: str


# =============================================================================
# SCHEMAS DE SUGERENCIAS
# =============================================================================

class SugerenciaRedaccion(BaseModel):
    """Sugerencia de redaccion para una seccion."""
    seccion_codigo: str
    campo: str
    texto_sugerido: str
    fuente: str  # corpus_sea, eia_aprobado, etc.
    confianza: float = Field(ge=0.0, le=1.0)


class SolicitudSugerenciaRequest(BaseModel):
    """Request para obtener sugerencias de redaccion."""
    capitulo_numero: int
    seccion_codigo: str
    campo: Optional[str] = None
    contexto_adicional: Optional[str] = None


class SugerenciasResponse(BaseModel):
    """Respuesta con sugerencias de redaccion."""
    sugerencias: List[SugerenciaRedaccion]
    total: int


# =============================================================================
# SCHEMAS DE INICIALIZACION
# =============================================================================

class IniciarCapituloRequest(BaseModel):
    """Request para iniciar recopilacion de un capitulo."""
    capitulo_numero: int


class IniciarCapituloResponse(BaseModel):
    """Respuesta al iniciar recopilacion de un capitulo."""
    capitulo_numero: int
    titulo: str
    secciones: List[SeccionConPreguntas]
    total_preguntas: int
    preguntas_obligatorias: int
    mensaje_bienvenida: str

File: frontend/src/types/recopilacion.ts
/**
 * Tipos TypeScript para Recopilacion Guiada (Fase 3).
 */

// =============================================================================
// ENUMS
// =============================================================================

export type EstadoSeccion = 'pendiente' | 'en_progreso' | 'completado' | 'validado'

export type TipoRespuesta =
  | 'texto'
  | 'numero'
  | 'fecha'
  | 'seleccion'
  | 'seleccion_multiple'
  | 'archivo'
  | 'coordenadas'
  | 'tabla'
  | 'booleano'

export type SeveridadInconsistencia = 'error' | 'warning' | 'info'

export type EstadoExtraccion = 'pendiente' | 'procesando' | 'completado' | 'error'

// =============================================================================
// CONSTANTES UI
// =============================================================================

export const ESTADO_SECCION_LABELS: Record<EstadoSeccion, string> = {
  pendiente: 'Pendiente',
  en_progreso: 'En Progreso',
  completado: 'Completado',
  validado: 'Validado',
}

export const ESTADO_SECCION_COLORS: Record<EstadoSeccion, string> = {
  pendiente: 'badge-ghost',
  en_progreso: 'badge-warning',
  completado: 'badge-success',
  validado: 'badge-info',
}

export const SEVERIDAD_COLORS: Record<SeveridadInconsistencia, string> = {
  error: 'alert-error',
  warning: 'alert-warning',
  info: 'alert-info',
}

// =============================================================================
// INTERFACES DE PREGUNTAS
// =============================================================================

export interface OpcionSeleccion {
  valor: string
  etiqueta: string
  descripcion?: string
}

export interface ValidacionPregunta {
  requerido?: boolean
  min?: number
  max?: number
  patron?: string
  mensaje_patron?: string
}

export interface PreguntaRecopilacion {
  id: number
  capitulo_numero: number
  seccion_codigo: string
  seccion_nombre?: string
  codigo_pregunta: string
  pregunta: string
  descripcion?: string
  tipo_respuesta: TipoRespuesta
  opciones?: OpcionSeleccion[]
  valor_por_defecto?: unknown
  validaciones?: ValidacionPregunta
  es_obligatoria: boolean
  condicion_activacion?: Record<string, unknown>
  orden: number
  activo?: boolean
}

export interface PreguntaConRespuesta extends PreguntaRecopilacion {
  respuesta_actual?: unknown
  es_valida: boolean
  mensaje_validacion?: string
}

// =============================================================================
// INTERFACES DE SECCIONES
// =============================================================================

export interface SeccionBase {
  capitulo_numero: number
  seccion_codigo: string
  seccion_nombre?: string
}

export interface SeccionConPreguntas extends SeccionBase {
  preguntas: PreguntaConRespuesta[]
  total_preguntas: number
  preguntas_obligatorias: number
  preguntas_respondidas: number
  progreso: number
  estado: EstadoSeccion
}

export interface SeccionResumen extends SeccionBase {
  estado: EstadoSeccion
  progreso: number
  tiene_inconsistencias: boolean
  documentos_vinculados: number
}

// =============================================================================
// INTERFACES DE CONTENIDO
// =============================================================================

export interface RespuestaIndividual {
  codigo_pregunta: string
  valor: unknown
}

export interface GuardarRespuestasRequest {
  respuestas: RespuestaIndividual[]
}

export interface ContenidoEIA {
  id: number
  proyecto_id: number
  capitulo_numero: number
  seccion_codigo: string
  seccion_nombre?: string
  contenido: Record<string, unknown>
  estado: EstadoSeccion
  progreso: number
  documentos_vinculados: number[]
  validaciones: Record<string, unknown>[]
  inconsistencias: Record<string, unknown>[]
  sugerencias: Record<string, unknown>[]
  notas?: string
  created_at?: string
  updated_at?: string
}

// =============================================================================
// INTERFACES DE CAPITULOS Y PROGRESO
// =============================================================================

export interface CapituloProgreso {
  numero: number
  titulo: string
  total_secciones: number
  secciones_completadas: number
  progreso: number
  estado: EstadoSeccion
  tiene_inconsistencias: boolean
  secciones: SeccionResumen[]
}

export interface ProgresoGeneralEIA {
  proyecto_id: number
  total_capitulos: number
  capitulos_completados: number
  progreso_general: number
  total_inconsistencias: number
  capitulos: CapituloProgreso[]
}

export interface IniciarCapituloResponse {
  capitulo_numero: number
  titulo: string
  secciones: SeccionConPreguntas[]
  total_preguntas: number
  preguntas_obligatorias: number
  mensaje_bienvenida: string
}

// =============================================================================
// INTERFACES DE INCONSISTENCIAS
// =============================================================================

export interface InconsistenciaDetectada {
  regla_codigo: string
  regla_nombre: string
  capitulo_origen: number
  seccion_origen: string
  campo_origen: string
  valor_origen?: unknown
  capitulo_destino: number
  seccion_destino: string
  campo_destino: string
  valor_destino?: unknown
  mensaje: string
  severidad: SeveridadInconsistencia
  fecha_deteccion?: string
}

export interface ValidacionConsistenciaResponse {
  proyecto_id: number
  total_reglas_evaluadas: number
  inconsistencias: InconsistenciaDetectada[]
  errores: number
  warnings: number
  es_consistente: boolean
}

// =============================================================================
// INTERFACES DE EXTRACCION
// =============================================================================

export interface MapeoSeccionSugerido {
  capitulo_numero: number
  seccion_codigo: string
  campo: string
  valor: unknown
  confianza: number
}

export interface ExtraccionDocumento {
  id: number
  proyecto_id: number
  documento_id: number
  tipo_documento?: string
  datos_extraidos: Record<string, unknown>
  mapeo_secciones: MapeoSeccionSugerido[]
  confianza_extraccion: number
  estado: EstadoExtraccion
  errores: Record<string, unknown>[]
  procesado_por: string
  created_at?: string
  updated_at?: string
}

export interface ExtraccionDocumentoRequest {
  documento_id: number
  tipo_documento?: string
  forzar_reprocesar?: boolean
}

// =============================================================================
// INTERFACES DE VINCULACION
// =============================================================================

export interface VincularDocumentoRequest {
  documento_id: number
  capitulo_numero: number
  seccion_codigo: string
}
